docker/manager.py

```diff
@@ -0,0 +1,821 @@
+        """
+        if isinstance(value, str):
+            if "=" not in value:
+                self._error(
+                    field, f"{value} should have format ARG=foobar",
+                )
+            else:
+                envars, _, _ = value.partition("=")
+                self._validate_env_vars(True, field, envars)
+        if isinstance(value, list):
+            for v in value:
+                self._check_with_args_format(field, v)
+
+    def _check_with_dists_defined(self, field, value):
+        if isinstance(value, list):
+            for v in value:
+                self._check_with_dists_defined(field, v)
+        else:
+            if value not in self.releases:
+                self._error(field, f"{field} is not defined under releases")
+
+    def _validate_env_vars(self, env_vars, field, value):
+        """
+        Validate if given is a environment variable.
+
+        The rule's arguments are validated against this schema:
+            {'type': 'boolean'}
+        """
+        if isinstance(value, str):
+            if env_vars and not value.isupper():
+                self._error(field, f"{value} cannot be parsed to envars.")
+        else:
+            self._error(field, f"{value} cannot be parsed as string.")
+
+    def _validate_supported_dists(self, supported_dists, field, value):
+        """
+        Validate if given is a supported OS.
+
+        The rule's arguments are validated against this schema:
+            {'type': 'boolean'}
+        """
+        if isinstance(value, str):
+            if supported_dists and value not in SUPPORTED_OS:
+                self._error(
+                    field,
+                    f"{value} is not defined in SUPPORTED_OS. If you are adding a new distros make "
+                    f"sure to add it to SUPPORTED_OS",
+                )
+        if isinstance(value, list):
+            for v in value:
+                self._validate_supported_dists(supported_dists, field, v)
+
+
+# noinspection PyPep8Naming
+class cached_property(property):
+    """Direct ports from bentoml/utils/__init__.py"""
+
+    class _Missing(object):
+        def __repr__(self):
+            return "no value"
+
+        def __reduce__(self):
+            return "_missing"
+
+    _missing = _Missing()
+
+    def __init__(self, func, name=None, doc=None):
+        super().__init__(doc)
+        self.__name__ = name or func.__name__
+        self.__module__ = func.__module__
+        self.__doc__ = doc or func.__doc__
+        self.func = func
+
+    def __set__(self, obj, value):
+        obj.__dict__[self.__name__] = value
+
+    def __get__(self, obj, types=None):
+        if obj is None:
+            return self
+        value = obj.__dict__.get(self.__name__, self._missing)
+        if value is self._missing:
+            value = self.func(obj)
+            obj.__dict__[self.__name__] = value
+        return value
+
+
+class TemplatesMixin(object):
```

Comment:
`TemplatesMixin` will handle our templates generation and also provide a `build_path` dictionary where:
  - `keys`: image tag
  - `values`: path to generated Dockerfile

We can later use this `build_path` to build our image with `docker-py`